### DISCUSSION TITLE
TUI for MiniDexed

### DISCUSSION THREAD

**Discussion by nhemsley** [probonopd/MiniDexed#727] [Ideas]
Hello,

I have recently found MiniDex, and as advertised it boots up and outputs sound from my keyboard (nice!).

I believe I have setup my midi control codes properly, however I do not have an LCD screen attached, so I cannot view the menu system. I may watch a tutorial, and guesstimate in the short term, but that is tricky.

However I'm also interested in adding a TUI, controlled via a minidex.ini setting (EnableTUI?).

I have downloaded the repo, and am going to look at it. However a few questions:

Would this be something that the devs would like to see?
Does Minidex expose a working TTY? Or is it just printf style debugging? 
Is there anything else I havent thought of that makes this infeasible?

The other thought is a web interface, allowing to view/select performances etc. Edit settings, set control codes via 'press midi key'. Keyboard presets would be very good. 

This would require write access to the SD card to persist changes, which I'm not sure is currently the default (I imagine that's feasible).

Any thoughts welcome. If a web interface is feasible and sounds better, that probably has more feature possibilities.

I can add a separate discussion / issue for web if that is deemed preferable

**Comment by probonopd** [comment]
Hi @nhemsley. MiniDexed is a [MIDI](https://en.wikipedia.org/wiki/MIDI) instrument. So it receives commands on the serial line and/or on USB via MIDI. So your software would need to send MIDI commands to MiniDexed. The command set is that of the DX7, but for each of the 8 tone generators you can use a different MIDI channel.

Any software or controller that can control a DX7 (using MIDI SYSEX) should be able to control MiniDexed, too.

MiniDexed does not send out logging over serial, since the serial port is used for MIDI. Logging goes out via HDMI (if enabled in the `minidexed.ini` file).

**Reply by nhemsley** [reply]
Thanks @probonopd.

Basically I think what I want/need is enhanced Debug logging, which outputs the menu items via printf (or whichever), to HDMI. I have HDMI output, but no LCD.

I have tried putting a printf here: 

https://github.com/probonopd/MiniDexed/blob/8c18e60e5133a56e7208c4ff051cc2ec80e0aaa1/src/userinterface.cpp#L300

and enabled LCD support, but I believe the code is detecting if LCD initializes properly (which it won't be). I'll add another code path to treat HDMI as a separate output device, and see what I can see.

Basically:
I want to be able to see output on my HDMI, just like if I have installed an LCD.

I had a raspberry PI sitting around, but no LCD at this point.

Thanks!

**Reply by probonopd** [reply]
As far as I know, HDMI output works whether you have an LCD attached or not. But you will see different messages on HDMI than what is shown on the display. A basic display is around 3 USD, so we are unlikely to put huge efforts into HDMI.

**Reply by nhemsley** [reply]
Ok, understood.

**Reply by nhemsley** [reply]
Yes, HDMI output works, and can output MIDI codes, (and something else)
